File: ios/CellularChat/Transport/TransportCoordinator.swift
import Foundation
import CellularChatCore

/// Sequential transport arbitration and upgrade (PROTOCOL_V2.md §4/§10).
///
/// The two high-power discovery mechanisms (Wi-Fi Aware, Nearby) are never run
/// in parallel: each is attempted in order with a bounded timeout, and BLE is
/// the always-available baseline. Once an authenticated high-throughput path
/// wins, losing transports stop.
@MainActor
final class TransportCoordinator {

    /// Preferred order from §4: Wi-Fi Aware → Nearby Connections → BLE.
    static let defaultOrder: [TransportKind] = [.wifiAware, .nearby, .ble]

    private(set) var active: PeerTransport?

    /// Try transports in `order`, returning the first that connects. Only one
    /// connect attempt is in flight at a time (no parallel discovery). Testable
    /// with fake transports.
    static func arbitrate(transports: [PeerTransport],
                          order: [TransportKind] = defaultOrder) async -> PeerTransport? {
        for kind in order {
            guard let transport = transports.first(where: { $0.kind == kind && $0.isAvailable })
            else { continue }
            if case .success = await transport.connect() {
                return transport
            }
        }
        return nil
    }

    func select(transports: [PeerTransport],
                order: [TransportKind] = defaultOrder) async -> PeerTransport? {
        let winner = await Self.arbitrate(transports: transports, order: order)
        if let winner {
            active = winner
        }
        return winner
    }

    /// Whether an available transport ranks strictly above the active one and is
    /// therefore worth a same-`sid` upgrade (§10). Order index: lower is better.
    func shouldUpgrade(to candidate: TransportKind,
                       order: [TransportKind] = defaultOrder) -> Bool {
        guard let active, candidate != active.kind,
              let ai = order.firstIndex(of: active.kind),
              let ci = order.firstIndex(of: candidate) else { return false }
        return ci < ai
    }

    func teardown(reason: ReasonCode) {
        active?.disconnect(reason: reason)
        active = nil
    }
}

File: android/app/src/main/java/com/cellularchat/app/transport/TransportCoordinator.kt
package com.cellularchat.app.transport

/**
 * Sequential transport arbitration (IMPLEMENTATION_PLAN.md §4, PROTOCOL_V2.md
 * §10). Candidates are attempted strictly in preference order — Wi-Fi Aware,
 * then Nearby Connections, then the mandatory BLE GATT — each with its own
 * measured timeout. The first candidate that reaches an authenticated link
 * wins; the losers are never started in parallel, so two high-power discovery
 * mechanisms never run at once.
 *
 * Wholly synchronous-friendly: a candidate may invoke its callback inline
 * (tests) or later from a radio callback (production). Re-entrancy is guarded so
 * an inline callback still advances the ladder exactly once per candidate.
 */
class TransportCoordinator(
    private val candidates: List<TransportCandidate>,
    private val timeoutForTag: (String) -> Long = { tag ->
        when (tag) {
            "aware" -> 4_000L
            "nearby" -> 4_000L
            else -> 0L // BLE is mandatory: no arbitration timeout.
        }
    },
) {
    sealed interface Result {
        data class Won(val tag: String, val transport: PeerTransport) : Result
        data object Exhausted : Result
    }

    private var index = 0
    private var active: TransportCandidate? = null
    private var finished = false
    private var onResult: ((Result) -> Unit)? = null

    fun arbitrate(onResult: (Result) -> Unit) {
        check(this.onResult == null) { "arbitration already in progress" }
        this.onResult = onResult
        index = 0
        finished = false
        advance()
    }

    /** Aborts arbitration and cancels any in-flight attempt. */
    fun cancel() {
        if (finished) return
        finished = true
        active?.cancel()
        active = null
        onResult = null
    }

    private fun advance() {
        while (index < candidates.size && !finished) {
            val candidate = candidates[index]
            index += 1
            if (!candidate.isAvailable()) continue
            attempt(candidate)
            return
        }
        if (!finished) finish(Result.Exhausted)
    }

    private fun attempt(candidate: TransportCandidate) {
        active = candidate
        var settled = false
        candidate.attempt(
            timeoutForTag(candidate.tag),
            object : TransportCandidate.AttemptCallback {
                override fun onConnected(transport: PeerTransport) {
                    if (settled || finished) return
                    settled = true
                    active = null
                    finish(Result.Won(candidate.tag, transport))
                }

                override fun onFailed() {
                    if (settled || finished) return
                    settled = true
                    active = null
                    advance()
                }
            },
        )
    }

    private fun finish(result: Result) {
        finished = true
        val callback = onResult
        onResult = null
        callback?.invoke(result)
    }
}
